index.ts

```diff
@@ -33,7 +33,10 @@ type PublicMethodWithTarget<
   TMethod extends ValueOf<MessengerMethods>
 > = WithTarget<PublicMethod<TMethod>>;
 
-export type MessengerMeta = browser.runtime.MessageSender;
+export interface MessengerMeta {
+  trace: browser.runtime.MessageSender[];
+}
```

Comment:
This starts extending the `meta` object with more information, if necessary. Let's start with the trace, which I think covers your "sender" and "immediate sender" information.

`trace[0]` will always be the source of the message since generally the latter steps are just forwarders.

The issue that I mentioned in https://github.com/pixiebrix/pixiebrix-extension/issues/1542 is that the `trace[0]` might be `undefined` if we allow local calls. This is currently not caught by types in pixiebrix-extension, but it can be configured that way (requires making the whole codebase a lil bit stricter)